protocol.md

```diff
@@ -726,6 +727,141 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Challenge
+
+With this extension, Clients can authenticate follow-up requests to a [Creation](#creation) or [Creation With Upload](#creation-with-upload) 
+`POST` request using a high-entropy cryptographic random shared secret - and challenges
+based thereon.
+
+If the Server supports this extension, it MUST add `challenge` to the `Tus-Extension` header.
+
+#### Headers
+
+##### Upload-Secret
+
+The `Upload-Secret` MUST be a high-entropy cryptographic random string, consisting
+of 48 to 256 characters from the [base64 alphabet](https://tools.ietf.org/html/rfc4648#section-4).
+That's 36 to 192 bytes encoded as base64.
```

Comment:
I think this part is to specific to be honest. 48 to 256 characters might be enough for now but how about in 10 or even 5 years? I think we should change this part to be something like "The `Upload-Secret` MUST be a high-entropy cryptographic random string. The server MAY chose to reject any `Upload-Secret` it finds insecure. If the `Upload-Secret` is rejected the server MUST respond with a 400 Bad Request". What do you think?

Reply:
Of course, a longer secret would seem more future-proof. But looking at the whole construct (and other attack vectors), I am not sure it would really improve things much to allow more than 256 characters.

A few thoughts that went into picking this range:
- not defining a maximum:
    - would require additional error handling on the client side (for rejections based on exceeding the server's supported length) and protocol support (to specify the allowed maximum length)
    - could lead to inefficiencies: the client might only learn via a `400 Bad Request` response *after* already having sent a large upload
    - could open a DoS attack vector by filling up the server's disk if the server allows and stores arbitrary-sized values
- 256 characters is the maximum that still fits into a MySQL `VARCHAR` (which I once learnt has performance benefits over an essentially open ended type like `TEXT`), which - while not my personal favorite - remains a widely used database
- the minimum of 48 characters establishes a base level of security and already fits more than one gen4 UUID
- the `Upload-Challenge` is defined as `SHA256` (8 bytes), which caps the number of possible values at 2 ^ 64, making a brute force attack against the `Upload-Challenge` itself look significantly more efficient than against the `Upload-Secret` with its 2 ^ 288 (for 36 bytes/48 chars) to 2 ^ 1.536 (for 192 bytes/256 chars) possible values.

So, until the hash result length surpasses 48 bytes, I think the attack vector to worry about remains the hashing algorithm.

Reply:
All valid arguments 👍 I think this one is probably the most important and why we should have a defined way of calculating the secret:

>could lead to inefficiencies: the client might only learn via a 400 Bad Request response after already having sent a large upload

Reply:
Good reasoning. I think it's better to define limits instead of leaving them undefined. This makes building implementations easier.